rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,202 @@
+---
+authors: STeve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD 152 - Database Automatic User Provisioning for MongoDB
+
+## Required approvers
+
+Engineering: @r0mant || @smallinsky
+Product: @klizhentas || @xinding33
+Security: @reedloden || @jentfoo
+
+## What
+
+This RFD discusses on how to expand Database Automatic Provisioning feature for
+MongoDB.
+
+## Why
+
+Automatic User Provisioning has been implemented for several SQL databases,
+including PostgreSQL and MySQL, with the basic design described in [RFD
+113](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md).
+
+Adding support for database user provisioning to MongoDB presents unique
+challenges due to differences in architecture compared to traditional SQL
+databases.
+
+This RFD aims to identify the challenges and provide solutions to address them.
+
+Since the differences in architecture between MongoDB Atlas and self-hosted
+MongoDB are significant, they will be discussed separately in this RFD.
+
+## MongoDB Atlas Details
+
+Automatic User Provisioning will NOT be supported for MongoDB Atlas with the
+reasons discussed below. This RFD should be updated if better solutions are
+found in future iterations.
+
+[Database
+Users](https://www.mongodb.com/docs/atlas/security-add-mongodb-users/) and
+[Custom Database
+Roles](https://www.mongodb.com/docs/atlas/security-add-mongodb-roles/) for
+MongoDB Atlas are managed at a Atlas project level.
+
+As a consequence, database users and roles are NOT modifiable through
+in-database connections. Instead, one can authenticate with Atlas using the
+Atlas SDK and use APIs to manage these database users and roles. Multiple
+deployment jobs will be created to update the MongoDB clusters in this project,
+upon successful APIs calls.
+
+In my personal testing on an Atlas project with a single MongoDB cluster, it
+takes 10~20 seconds for the deployment job to refresh the database user in the
+target MongoDB cluster.
+
+With the current design of Automatic User Provisioning, the database user must
```

Comment:
This would only be a problem with the `drop` mode, right? With the `keep` mode the first connection will be slow but future connections will reuse the created user?

Reply:
Deployments are also done when changing database user's roles. For `keep` mode, if we are okay with NOT revoking the roles after a session (no role change, no "locking" the user account), from security perspective.

Reply:
I think we would still need to verify the permissions on connect.  Which I assume would only introduce a delay if the role was updated.

Reply:
@jentfoo 

> I think we would still need to verify the permissions on connect. Which I assume would only introduce a delay if the role was updated.

We should definitely check if the user's current role matches what we want to set. So we are good if not revoking the roles or locking the user when db sessions end?

Reply:
> So we are good if not revoking the roles or locking the user when db sessions end?

Unclear to me, I would need to understand more about how we plan to add Atlas support.  I think there are concerns around revoking access to Atlas as it extends the functionality (for example the Triggers functionality mentioned above).  We would likely need to revoke access proactively if it is revoked within Teleport.

Feel free to DM me with an outline of what you were thinking to support Atlas and we can brainstorm it.

Reply:
@jentfoo my current idea to support Atlas is pretty straight forward, 
- Teleport does the user/role setup on first connection (it may just timed out). 
- We do nothing when auto-provisioned db sessions ends.
- On next db sessions, we check if the database user has the correct roles. If yes, then nothing to do (no need Atlas deployment)

The concern is when there are NO active database sessions, the database user is still granted those roles and the database user is not "locked". Someone can still log in as that database user (not through Teleport auto-user provisioning flow).

Reply:
That's ok, we can keep Atlas out of scope for now.